File: rust/crates/engine/tests/engine_tests.rs
//! Port of `tests/test_engine.py`. Engine-level tests using the mock
//! backend + local cache source — no real OS or hardware required. Covers
//! the two safety-critical behaviors called out in docs/Architecture.md
//! section 3.3: the install batch must abort if the restore point fails,
//! and unconfirmed (ambiguous/upgrade-tier) entries must never be applied
//! silently.

use std::collections::{HashMap, HashSet};

use tempfile::tempdir;
use time::macros::date;

use waypoint_core::{Device, InstalledDriver, SignatureType};
use waypoint_engine::{AuditLog, WaypointEngine};
use waypoint_platform::MockDeviceBackend;
use waypoint_sources::LocalCacheSource;

fn cache_source(root: &std::path::Path) -> LocalCacheSource {
    let cache_dir = root.join("cache");
    let source = LocalCacheSource::new(&cache_dir).expect("create cache source");
    let driver_file = root.join("driver.inf");
    std::fs::write(&driver_file, "; fake inf for test purposes").unwrap();
    source
        .add_package(
            &driver_file,
            "HWID1",
            "{class}",
            "2.0",
            Some(date!(2026 - 01 - 01)),
            "Acme",
            SignatureType::Whql,
        )
        .expect("add_package");
    source
}

fn fake_gpu(problem_code: Option<i32>, installed: Option<InstalledDriver>) -> Device {
    Device {
        hwids: vec!["HWID1".to_string()],
        class_guid: "{class}".to_string(),
        class_name: "Display".to_string(),
        friendly_name: "Fake GPU".to_string(),
        instance_id: "DEV1".to_string(),
        problem_code,
        installed,
    }
}

#[test]
fn test_apply_aborts_when_restore_point_fails() {
    let tmp = tempdir().unwrap();
    let source = cache_source(tmp.path());
    let audit = AuditLog::new(tmp.path().join("audit.jsonl")).unwrap();

    let device = fake_gpu(Some(28), None);
    let mut backend = MockDeviceBackend::new(vec![device.clone()]);
    backend.restore_point_should_succeed = false;

    let engine = WaypointEngine::new(
        Box::new(backend),
        vec![Box::new(source)],
        audit,
        SignatureType::Attestation,
    );
    let assessments = engine.scan().unwrap();
    let plan = engine.build_plan(&assessments).unwrap();

    let devices_by_id = HashMap::from([("DEV1".to_string(), device)]);
    let result = engine.apply(
        &plan,
        &devices_by_id,
        false,
        tmp.path().join("backups").to_str().unwrap(),
        &HashSet::new(),
    );
    assert!(result.is_err());
    assert!(result.unwrap_err().contains("Restore point"));

    // The Python test also asserts `backend.install_calls == []` here.
    // `engine.backend` is a `Box<dyn DeviceBackend>` owned by `engine`, so
    // there's no external handle left to inspect after construction.
    // Structurally this is still verified: `apply()` returns the restore-
    // point error before its per-entry loop begins (see session.rs), so no
    // `install_driver` call can have happened on this path.
}

#[test]
fn test_missing_device_installs_without_confirmation_required() {
    let tmp = tempdir().unwrap();
    let source = cache_source(tmp.path());
    let audit = AuditLog::new(tmp.path().join("audit.jsonl")).unwrap();

    let device = fake_gpu(Some(28), None);
    let backend = MockDeviceBackend::new(vec![device.clone()]);

    let engine = WaypointEngine::new(
        Box::new(backend),
        vec![Box::new(source)],
        audit,
        SignatureType::Attestation,
    );
    let assessments = engine.scan().unwrap();
    let plan = engine.build_plan(&assessments).unwrap();
    // "missing" tier is not gated.
    assert!(!plan.entries[0].requires_confirmation);

    let devices_by_id = HashMap::from([("DEV1".to_string(), device)]);
    let results = engine
        .apply(
            &plan,
            &devices_by_id,
            true,
            tmp.path().join("backups").to_str().unwrap(),
            &HashSet::new(),
        )
        .unwrap();
    assert_eq!(results.len(), 1);
    assert!(results[0].success);
}

#[test]
fn test_upgrade_tier_is_skipped_without_explicit_confirmation() {
    let tmp = tempdir().unwrap();
    let source = cache_source(tmp.path());
    let audit = AuditLog::new(tmp.path().join("audit.jsonl")).unwrap();

    let installed = InstalledDriver {
        version: "1.0".to_string(),
        driver_date: Some(date!(2024 - 01 - 01)),
        publisher: "Acme".to_string(),
        signature_type: SignatureType::Whql,
        inf_path: None,
    };
    let device = fake_gpu(None, Some(installed));
    let backend = MockDeviceBackend::new(vec![device.clone()]);

    let engine = WaypointEngine::new(
        Box::new(backend),
        vec![Box::new(source)],
        audit,
        SignatureType::Attestation,
    );
    let assessments = engine.scan().unwrap();
    let plan = engine.build_plan(&assessments).unwrap();
    // upgrade tier is gated by default
    assert!(plan.entries[0].requires_confirmation);

    let devices_by_id = HashMap::from([("DEV1".to_string(), device)]);
    let backup_dir = tmp.path().join("backups").to_str().unwrap().to_string();

    let results = engine
        .apply(&plan, &devices_by_id, true, &backup_dir, &HashSet::new())
        .unwrap();
    assert!(results.is_empty()); // skipped: not confirmed

    let confirmed = HashSet::from(["DEV1".to_string()]);
    let results_confirmed = engine
        .apply(&plan, &devices_by_id, true, &backup_dir, &confirmed)
        .unwrap();
    assert_eq!(results_confirmed.len(), 1);
}

File: dotnet/Waypoint.Engine.Tests/EngineTests.cs
// Engine tests on the mock backend — no real OS or hardware. Covers the
// safety-critical behavior in docs/Architecture.md section 3.3: the batch
// aborts if the restore point fails, and unconfirmed entries never apply.

using System.Text.Json;
using Waypoint.Core;
using Waypoint.Platform;

namespace Waypoint.Engine.Tests;

public class EngineTests : IDisposable
{
    private readonly string _tempDir = Path.Combine(
        Path.GetTempPath(),
        "waypoint-engine-tests",
        Guid.NewGuid().ToString("N"));

    private readonly AuditLog _audit;

    public EngineTests()
    {
        Directory.CreateDirectory(_tempDir);
        _audit = new AuditLog(Path.Combine(_tempDir, "audit.jsonl"));
    }

    public void Dispose()
    {
        GC.SuppressFinalize(this);
        try
        {
            Directory.Delete(_tempDir, recursive: true);
        }
        catch (IOException)
        {
            // A leftover temp dir is not a test failure.
        }
        catch (UnauthorizedAccessException)
        {
        }
    }

    [Fact]
    public void Apply_AbortsWhenRestorePointFails_AndInstallsNothing()
    {
        var device = MissingDevice();
        var backend = new MockDeviceBackend([device]) { RestorePointShouldSucceed = false };
        var engine = NewEngine(backend);

        var plan = engine.BuildPlan(engine.Scan());

        var error = Assert.Throws<InvalidOperationException>(() => engine.Apply(
            plan,
            new Dictionary<string, Device> { ["DEV1"] = device },
            dryRun: false,
            backupDir: BackupDir));

        Assert.Contains("Restore point", error.Message);
        Assert.Empty(backend.InstallCalls);
        Assert.Contains(_audit.ReadAll(), r => r["event"].GetString() == "apply_aborted");
    }

    [Fact]
    public void MissingDevice_InstallsWithoutConfirmationRequired()
    {
        var device = MissingDevice();
        var backend = new MockDeviceBackend([device]);
        var engine = NewEngine(backend);

        var plan = engine.BuildPlan(engine.Scan());
        Assert.False(plan.Entries[0].RequiresConfirmation);

        var results = engine.Apply(
            plan,
            new Dictionary<string, Device> { ["DEV1"] = device },
            dryRun: true,
            backupDir: BackupDir);

        Assert.Single(results);
        Assert.True(results[0].Success);
    }

    [Fact]
    public void UpgradeTier_IsSkippedWithoutExplicitConfirmation()
    {
        var device = UpgradableDevice();
        var backend = new MockDeviceBackend([device]);
        var engine = NewEngine(backend);
        var devices = new Dictionary<string, Device> { ["DEV1"] = device };

        var plan = engine.BuildPlan(engine.Scan());
        Assert.True(plan.Entries[0].RequiresConfirmation);

        Assert.Empty(engine.Apply(plan, devices, dryRun: true, backupDir: BackupDir));
        Assert.Contains(_audit.ReadAll(), r => r["event"].GetString() == "skip_unconfirmed");

        var confirmed = engine.Apply(
            plan,
            devices,
            dryRun: true,
            backupDir: BackupDir,
            confirmedInstanceIds: new HashSet<string> { "DEV1" });

        Assert.Single(confirmed);
    }

    [Fact]
    public void RestorePointIsCreated_BeforeAnyBackupOrInstall()
    {
        var device = UpgradableDevice();
        var backend = new SequenceRecordingBackend(device);
        var engine = NewEngine(backend);

        var plan = engine.BuildPlan(engine.Scan());
        engine.Apply(
            plan,
            new Dictionary<string, Device> { ["DEV1"] = device },
            dryRun: false,
            backupDir: BackupDir,
            confirmedInstanceIds: new HashSet<string> { "DEV1" });

        Assert.Equal(["restore_point", "backup:DEV1", "install:DEV1"], backend.Sequence);
    }

    [Fact]
    public void AmbiguousDevice_IsGated_EvenWhenItsStatusIsNotUpgradeAvailable()
    {
        // Two devices claiming the same HWID: both "missing", both ambiguous.
        var first = MissingDevice();
        var second = MissingDevice() with { InstanceId = "DEV2", FriendlyName = "Fake GPU 2" };
        var backend = new MockDeviceBackend([first, second]);
        var engine = NewEngine(backend);
        var devices = new Dictionary<string, Device> { ["DEV1"] = first, ["DEV2"] = second };

        var plan = engine.BuildPlan(engine.Scan());
        Assert.All(plan.Entries, e => Assert.Equal(DeviceStatus.Missing.ToWireString(), e.Status));
        Assert.All(plan.Entries, e => Assert.True(e.RequiresConfirmation));

        Assert.Empty(engine.Apply(plan, devices, dryRun: true, backupDir: BackupDir));

        var confirmed = engine.Apply(
            plan,
            devices,
            dryRun: true,
            backupDir: BackupDir,
            confirmedInstanceIds: new HashSet<string> { "DEV1" });

        Assert.Equal("DEV1", Assert.Single(confirmed).InstanceId);
    }

    [Fact]
    public void AuditLog_ReadsBackRecords_SkipsBlankLines_AndToleratesAMissingFile()
    {
        var path = Path.Combine(_tempDir, "nested", "fresh.jsonl");
        var log = new AuditLog(path);
        Assert.Empty(log.ReadAll());

        log.Record("install", ("instance_id", "PCI\\VEN_10DE&DEV_2504"), ("dry_run", true), ("commands", new[] { "pnputil" }));
        File.AppendAllText(path, "\n   \n");

        var record = Assert.Single(log.ReadAll());
        Assert.Equal("install", record["event"].GetString());
        // '&' stays literal, as Python's json.dumps leaves it.
        Assert.Equal("PCI\\VEN_10DE&DEV_2504", record["instance_id"].GetString());
        Assert.Equal(JsonValueKind.True, record["dry_run"].ValueKind);
        Assert.Equal(["pnputil"], record["commands"].EnumerateArray().Select(e => e.GetString()));
        Assert.StartsWith(DateTime.UtcNow.ToString("yyyy-MM-dd"), record["timestamp"].GetString());
    }

    [Fact]
    public void PlanJson_KeepsThePythonCliShape()
    {
        var matched = MissingDevice();
        var unmatched = new Device(["HWID_NO_CANDIDATE"], "{class}", "Net", "Fake NIC", "DEV3", ProblemCode: 28);
        var engine = NewEngine(new MockDeviceBackend([matched, unmatched]));

        var plan = engine.BuildPlan(engine.Scan());

        using var document = JsonDocument.Parse(plan.ToJson());
        var entries = document.RootElement.GetProperty("entries").EnumerateArray().ToList();
        Assert.Equal(2, entries.Count);
        Assert.Equal(
            ["instance_id", "friendly_name", "status", "chosen_candidate", "ambiguous", "requires_confirmation"],
            entries[0].EnumerateObject().Select(p => p.Name));
        Assert.Equal("missing", entries[0].GetProperty("status").GetString());
        Assert.Equal(JsonValueKind.Null, entries[1].GetProperty("chosen_candidate").ValueKind);

        var candidate = entries[0].GetProperty("chosen_candidate");
        Assert.Equal("whql", candidate.GetProperty("signature_type").GetString());
        Assert.Equal("2026-01-01", candidate.GetProperty("driver_date").GetString());
        Assert.Equal(1024, candidate.GetProperty("size_bytes").GetInt64());
    }

    private string BackupDir => Path.Combine(_tempDir, "backups");

    private WaypointEngine NewEngine(IDeviceBackend backend) =>
        new(backend, [new FakeSource(Candidate)], _audit);

    private static readonly DriverCandidate Candidate = new(
        Hwid: "HWID1",
        ClassGuid: "{class}",
        Version: "2.0",
        DriverDate: new DateOnly(2026, 1, 1),
        Publisher: "Acme",
        SignatureType: SignatureType.Whql,
        Sha256: "abc123",
        SizeBytes: 1024,
        SourceId: "fake",
        SourceUrl: "https://example.test",
        DownloadUri: "/cache/blobs/abc123/driver.inf");

    // Problem code 28 = no driver bound.
    private static Device MissingDevice() =>
        new(["HWID1"], "{class}", "Display", "Fake GPU", "DEV1", ProblemCode: 28);

    private static Device UpgradableDevice() =>
        new(
            ["HWID1"],
            "{class}",
            "Display",
            "Fake GPU",
            "DEV1",
            Installed: new InstalledDriver("1.0", new DateOnly(2024, 1, 1), "Acme", SignatureType.Whql));

    // Stands in for LocalCacheSource, which lands in its own slice.
    private sealed class FakeSource(params DriverCandidate[] candidates) : IDriverSource
    {
        public string SourceId => "fake";

        public IReadOnlyList<DriverCandidate> Search(IReadOnlyList<string> hwids) =>
            candidates.Where(c => hwids.Contains(c.Hwid)).ToList();

        public Task<string> FetchAsync(DriverCandidate candidate, string destDir, CancellationToken cancellationToken = default) =>
            Task.FromResult(candidate.DownloadUri);
    }

    // Records call order, which MockDeviceBackend's per-call lists cannot show.
    private sealed class SequenceRecordingBackend(params Device[] devices) : IDeviceBackend
    {
        public List<string> Sequence { get; } = [];

        public IReadOnlyList<Device> EnumerateDevices() => devices;

        public string ExportDriverBackup(Device device, string destDir)
        {
            Sequence.Add($"backup:{device.InstanceId}");
            return $"{destDir}/{device.InstanceId}.backup";
        }

        public InstallResult InstallDriver(Device device, string packagePath, bool dryRun)
        {
            Sequence.Add($"install:{device.InstanceId}");
            return new InstallResult(true, [], "installed");
        }

        public bool CreateRestorePoint(string description)
        {
            Sequence.Add("restore_point");
            return true;
        }
    }
}
